### PR TITLE
Make the source actor more responsive

### PR DESCRIPTION
**Pull request by tobim**
TODO:
- [x] Changelog entry.

### CODE REVIEW

File: libvast/vast/format/reader.hpp [hunk 1]
```diff
@@ -110,7 +110,7 @@ class reader {
 
   caf::atom_value table_slice_type_;
   std::chrono::steady_clock::duration read_timeout_
-    = vast::defaults::import::read_timeout;
+    = std::chrono::milliseconds{20};
```
Comment (tobim) [inline libvast/vast/format/reader.hpp:113]: Let's agree on a sensible value here before touching `defaults.hpp`.

File: libvast/vast/system/source.hpp [hunk 2]
```diff
@@ -257,46 +260,26 @@ source(caf::stateful_actor<source_state<Reader>>* self, Reader reader,
...
-        } else {
-          if (err != vast::ec::end_of_input)
-            VAST_INFO(self, "completed with message:", render(err));
-          return finish();
+      if (err == ec::timeout) {
+        if (!st.waiting_for_input) {
+          // This pull handler was invoked while we were waiting for a wakeup
+          // message. Sending another one would create a parallel wakeup cycle.
+          st.waiting_for_input = true;
+          self->delayed_send(self, std::chrono::seconds{1}, atom::wakeup_v);
```
Comment (dominiklohmann) [inline libvast/vast/system/source.hpp:275]: Shouldn't this just be a normal send to enqueue the message to the back of the message queue?

Comment (tobim) [reply libvast/vast/system/source.hpp]: We agreed to implement an exponential backoff mechanism here.

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,18 @@ Every entry has a category for which we use the following visual abbreviations:
...
+- ⚠️ The `read-timeout` option has been added to allow setting the input
+  timeout for low volume data sources. When this timeout is reached while
+  a source is waiting for input, the current batch will be closed sent to
+  the server. This behavior was previously controlled by `batch-timeout`,
+  which still exists and controls the maximum buffer time before batches
+  are sent to the server.
+  [#1096](https://github.com/tenzir/vast/pull/1096)
+
+- 🐞 Input starved sources aren't going to block calls to `vast status` any
+  longer.
```
Comment (dominiklohmann) [inline CHANGELOG.md:23]:
```suggestion
- 🐞 Sources that receive no or very little input do not block `vast status`
  any longer.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,18 @@ Every entry has a category for which we use the following visual abbreviations:
 
 ## Unreleased
 
+- ⚠️ The `read-timeout` option has been added to allow setting the input
+  timeout for low volume data sources. When this timeout is reached while
+  a source is waiting for input, the current batch will be closed sent to
+  the server. This behavior was previously controlled by `batch-timeout`,
+  which still exists and controls the maximum buffer time before batches
+  are sent to the server.
```
Comment (dominiklohmann) [inline CHANGELOG.md:19]:
```suggestion
- ⚠️ The new option `import.read-timeout` allows for setting an input timeout
  for low volume sources. Reaching the timeout causes the current batch to be
  forwarded immediately. This behavior was previously controlled by
  `import.batch-timeout`, which now only controls the maximum buffer time
  before the source forwards batches to the server.
```

File: doc/cli/vast-import.md [hunk 5]
```diff
@@ -82,6 +82,13 @@ size.
...
+Sets a time limit for forwarding buffered table slices to the importer. If the
+time limit is hit while the source is still busy processing data, the current
+batch will be finalized and all buffered batches will be send to the node. The
+default batch timeout is 1 second.
+
+#### `vast.import.read-timeout`
+
+Sets a timeout for reading from input sources. If no data is received for the
+set amount of time, the process will close the current batch and retry at a
+later time. The default read timeout is 20 milliseconds.
```
Comment (dominiklohmann) [inline doc/cli/vast-import.md:94]: Can you rewrite both of these to use active voice?

File: libvast/src/format/csv.cpp [hunk 6]
```diff
@@ -424,38 +416,36 @@ caf::error reader::read_impl(size_t max_events, size_t max_slice_size,
...
-      VAST_DEBUG(this, "reached input timeout at line", lines_->line_number());
-      return finish(callback, ec::timeout);
+    if (batch_timeout_ > decltype(batch_timeout_)::zero()
+        && start + batch_timeout_ < std::chrono::steady_clock::now()) {
+      VAST_DEBUG(this, "reached input time limit");
+      break;
     }
+    bool timed_out = next_line();
+    if (timed_out)
+      return ec::timeout;
```
Comment (dominiklohmann) [inline libvast/src/format/csv.cpp:448]: Does it make sense to invert the return value of `next_line()`, such that we can write this instead?

```cpp
if (!next_line())
  return ec::timeout;
```

File: libvast/src/format/zeek.cpp [hunk 7]
```diff
@@ -276,13 +266,21 @@ caf::error reader::read_impl(size_t max_events, size_t max_slice_size,
   // Sanity checks.
   VAST_ASSERT(max_events > 0);
   VAST_ASSERT(max_slice_size > 0);
+  auto next_line = [&] {
+    auto timed_out = lines_->next_timeout(read_timeout_);
+    if (timed_out)
+      VAST_DEBUG(this, "reached input timeout at line", lines_->line_number());
+    return timed_out;
+  };
```
Comment (dominiklohmann) [inline libvast/src/format/zeek.cpp:274]: Should this be a member function on the line range instead to deduplicate it?

File: libvast/src/system/application.cpp [hunk 8]
```diff
@@ -273,10 +273,10 @@ auto make_spawn_source_command() {
       .add<caf::atom_value>("batch-encoding", "encoding type of table slices "
                                               "(arrow or msgpack)")
       .add<size_t>("batch-size", "upper bound for the size of a table slice")
-      .add<std::string>("batch-timeout", "timoeut after which batched table "
-                                         "slices are forwarded")
-      .add<size_t>("max-events,n", "the maximum number of events to "
-                                   "import"));
+      .add<std::string>("batch-timelimit", "time limit after which batched "
+                                           "table slices are forwarded")
```
Comment (dominiklohmann) [inline libvast/src/system/application.cpp:277]: Why did you rename this from timeout to timelimit? The latter is a made up word.

```suggestion
      .add<std::string>("batch-timeout", "timeout after which batched table "
                                         "slices are forwarded")
```

Comment (dominiklohmann) [reply libvast/src/system/application.cpp:277]: Note that this must be applied all throughout the PR.

File: libvast_test/src/events.cpp [hunk 9]
```diff
@@ -89,9 +89,11 @@ extract(Reader&& reader, table_slice::size_type slice_size) {
 template <class Reader>
 std::vector<table_slice_ptr>
 inhale(const char* filename, table_slice::size_type slice_size) {
+  caf::settings settings;
+  // A non-positive value disables the time limit.
```
Comment (dominiklohmann) [inline libvast_test/src/events.cpp:93]: Can you add to the comment that we want this because of the weird clock in the test setup?

File: vast.yaml.example [hunk 10]
```diff
@@ -143,15 +143,15 @@ vast:
     # The maximum number of events to import.
     #max-events: <infinity>
     # Timeout after which batched table slices are forwarded.
-    batch-timeout: 10s
+    read-timeout: 10s
```
Comment (dominiklohmann) [inline vast.yaml.example:146]: Can you add the new option to the example config as well?

**Review by dominiklohmann — CHANGES_REQUESTED** [review]
This works quite well, and I am happy with the new default values. Just some comments that are not related to the functionality.

File: libvast/src/system/application.cpp (hunk 8 above)
Comment (tobim) [reply libvast/src/system/application.cpp]: I missed those sites. `batch-timelimit` was the first naming approach when I split the functionality out from `batch-timeout`.

File: libvast/src/format/csv.cpp (hunk 6 above)
Comment (tobim) [reply libvast/src/format/csv.cpp:448]: The next person reading that code will probably not have the semantics of `next_line()` present. Making the meaning of the return value explicit by naming it `timed_out` helps imho.

File: libvast/src/format/zeek.cpp (hunk 7 above)
Comment (tobim) [reply libvast/src/format/zeek.cpp:274]: The lambda only exists so the log statement is not duplicated. If we move it to the line range the message would read `vast.detail.line_range reached ...` instead of `vast.format.zeek.reader reached ...`.

Comment (dominiklohmann) [reply libvast/src/format/zeek.cpp:274]: Fair point.

File: doc/cli/vast-import.md [hunk 11]
```diff
@@ -82,6 +82,17 @@ size.
...
 #### `vast.import.batch-timeout`
 
-Sets a timeout for forwarding buffered table slices to the importer. If the
-timeout fires before a table slice reaches `vast.import.batch-size`, then the
-table slice will contain fewer events but ship immediately.
+Sets a timeout for forwarding buffered table slices to the importer.
+
+The `vast.import.batch-timeout` option controls the maximum buffering period
+until table slices are forwarded to the node. The default batch timeout is 1
+second.
```
Comment (dominiklohmann) [inline doc/cli/vast-import.md:89]:
```suggestion
until table slices are forwarded to the node. The default batch timeout is
one second.
```

File: doc/cli/vast-import.md [hunk 12]
```diff
@@ -82,6 +82,17 @@ size.
...
+
+#### `vast.import.read-timeout`
+
+Sets a timeout for reading from input sources.
+
+The `vast.import.read-timeout` option controls the time that a call to read
+data from the input will block.  If no data is received for the set amount of
+time, the process will yield and try again at a later time. The default read
+timeout is 20 milliseconds.
```
Comment (dominiklohmann) [inline doc/cli/vast-import.md:98]: Passive to active voice.

File: vast.yaml.example [hunk 13]
```diff
@@ -138,23 +138,28 @@ vast:
...
-    # Timeout after which batched table slices are forwarded.
-    batch-timeout: 10s
-    # Read timoeut after which data is forwarded to the importer regardless of
-    # batching and table slices being unfinished.
-    read-timeout: 10s
-    # Block until the importer forwarded all data.
-    blocking: false
+    # Timeout after which buffered table slices are forwarded to the node.
+    # Currently unfinished table slices will be finished and forwarded when
+    # this timeout is hit.
```
Comment (dominiklohmann) [inline vast.yaml.example:148]: Passive to active voice and future to present tense.

File: vast.yaml.example [hunk 14]
```diff
@@ -138,23 +138,28 @@ vast:
...
+    # vast.import.read-timeout option only.
     batch-size: 100
     # Encoding type of table slices (arrow or msgpack).
     batch-encoding: arrow
+    # Block until the importer forwarded all data.
+    blocking: false
+    # Timeout the for read function. Input streams are polled before data is
+    # read from them. This mechansim helps with keeping the import process
+    # responsive. The amount of time that each iteration waits for new input
+    # can be configured with this option.
```
Comment (dominiklohmann) [inline vast.yaml.example:161]: Passive to active voice.

**Review by dominiklohmann — CHANGES_REQUESTED** [review]
Just wording now.

File: libvast/vast/system/source.hpp [hunk 15]
```diff
@@ -257,46 +263,35 @@ source(caf::stateful_actor<source_state<Reader>>* self, Reader reader,
...
+          // This pull handler was invoked while we were waiting for a wakeup
+          // message. Sending another one would create a parallel wakeup cycle.
+          st.waiting_for_input = true;
+          self->delayed_send(self, st.wakeup_delay, atom::wakeup_v);
+          // Exponential backoff for the wakeup calls.
+          // For each consecutive invocation of this generate handler that does
+          // not emit any events, we double the wakeup delay.
+          // The sequence is 0, 20, 40, 80, 160, 320, 640, 1280.
+          if (st.wakeup_delay == std::chrono::milliseconds::zero())
+            st.wakeup_delay = std::chrono::milliseconds{20};
```
Comment (dominiklohmann) [inline libvast/vast/system/source.hpp:284]: Is it just coincidence that we're using the same initial value here as the read timeout, or should this be configurable via `import.read-timeout` as well?

I personally think it is fine to leave these fixed.

File: doc/cli/vast-import.md [hunk 16]
```diff
@@ -82,6 +82,17 @@ size.
...
+second.
+
+#### `vast.import.read-timeout`
+
+Sets a timeout for reading from input sources.
+
+The `vast.import.read-timeout` option controls the time that a call to read data
+from the input will block. The process yields and tries again at a later time if
+no data is received for the set value. The default read timeout is 20
+milliseconds.
```
Comment (dominiklohmann) [inline doc/cli/vast-import.md:98]: The first sentence does not read very fluently.

File: libvast/vast/format/reader.hpp (hunk 1 above)
Comment (dominiklohmann) [reply libvast/vast/format/reader.hpp:116]: I'm fine with the value, but this should still be changed to use the value set in `defaults.hpp`.

File: libvast/vast/system/source.hpp (hunk 15 above)
Comment (tobim) [reply libvast/vast/system/source.hpp:284]: It is coincidence. But we should cap the maximum at `batch-timeout / 2`.

Comment (dominiklohmann) [reply libvast/vast/system/source.hpp:284]: That is a sensible choice.

File: libvast/src/format/csv.cpp (hunk 6 above)
Comment (dominiklohmann) [reply libvast/src/format/csv.cpp:448]: I don't have too strong of an opinion on this, so feel free to just mark as resolved.

File: libvast/vast/system/source.hpp (hunk 15 above)
Comment (tobim) [reply libvast/vast/system/source.hpp:284]: This is done now.

File: doc/cli/vast-import.md (hunk 16 above)
Comment (tobim) [reply doc/cli/vast-import.md]: Updated to "determines how long a call to read data from the input will block".

File: libvast/vast/defaults.hpp [hunk 17]
```diff
@@ -57,7 +57,7 @@ constexpr size_t max_events = 0;
 
 /// Read timoeut after which data is forwarded to the importer regardless of
 /// batching and table slices being unfinished.
-constexpr std::chrono::milliseconds read_timeout = std::chrono::seconds{10};
+constexpr std::chrono::milliseconds batch_timeout = std::chrono::seconds{10};
```
Comment (dominiklohmann) [inline libvast/vast/defaults.hpp:60]: The comment needs to be updated and both `read_timeout` and `batch_timeout` should have their default set here.

File: libvast/vast/format/reader.hpp [hunk 18]
```diff
@@ -108,9 +112,13 @@ class reader {
   virtual caf::error read_impl(size_t max_events, size_t max_slice_size,
                                consumer& f) = 0;
 
+public:
   caf::atom_value table_slice_type_;
-  std::chrono::steady_clock::duration read_timeout_
-    = vast::defaults::import::read_timeout;
+  reader_clock::duration read_timeout_ = std::chrono::milliseconds{20};
```
Comment (dominiklohmann) [inline libvast/vast/format/reader.hpp:117]:
```suggestion
  reader_clock::duration read_timeout_ = vast::defaults::import::read_timeout;
```

**Review by dominiklohmann — CHANGES_REQUESTED** [review]

**Review by dominiklohmann — APPROVED** [review]
Nice work! The CI failure is unrelated. PR should be rebased once that is fixed.
